Content.Server/Chat/Systems/ChatSystem.Events.cs

```diff
@@ -0,0 +1,211 @@
+        /// </remarks>
+        public Type? ClaimedBy;
+
+        public EntityChat(EntityUid source, string message)
+        {
+            Source = source;
+            Message = message;
+        }
+    }
+
+    // New Chat flow:
+    //
+    // Parse                    - we have no idea what we have
+    // Attempt                  - now we know what type of chat this is, do we continue?
+    // GetRecipients            - we get who will receive this chat
+    // Transform                - we can change what the speaker says
+    //
+    // BeforeEntityChat         - we could cancel the chat at the recipient
+    // GotEntityChatTransform   - we can change what the recipient hears
+    // GotEntityChat            - finally, show the user the message
+
+    [ByRefEvent]
+    public struct EntityChatParseEvent
+    {
+        public EntityChat Chat;
+
+        public EntityChatParseEvent(EntityChat chat)
+        {
+            Chat = chat;
+        }
+
+        public bool Handled;
+    }
+
+    [ByRefEvent]
+    public struct EntityChatAttemptEvent
+    {
+        public EntityChat Chat;
+
+        public EntityChatAttemptEvent(EntityChat chat)
+        {
+            Chat = chat;
+        }
+
+        public bool Cancelled { get; private set; }
+        public void Cancel() => Cancelled = true;
+        public void Uncancel() => Cancelled = false;
+    }
+
+    [ByRefEvent]
+    public struct EntityChatGetRecipientsEvent
+    {
+        public readonly EntityChat Chat;
+
+        public EntityChatGetRecipientsEvent(EntityChat chat)
+        {
+            Chat = chat;
+        }
+
+        public bool Handled;
+    }
+
+    [ByRefEvent]
+    public struct EntityChatTransformEvent
+    {
+        public EntityChat Chat;
+
+        public EntityChatTransformEvent(EntityChat chat)
+        {
+            Chat = chat;
+        }
+
+        public bool Handled;
+    }
+
+    [ByRefEvent]
+    public struct BeforeEntityChatEvent
+    {
+        public readonly EntityUid Recipient;
+        public readonly EntityChat Chat;
+
+        public BeforeEntityChatEvent(EntityUid recipient, EntityChat chat)
+        {
+            Recipient = recipient;
+            Chat = chat;
+        }
+
+        public bool Cancelled { get; private set; }
+        public void Cancel() => Cancelled = true;
+        public void Uncancel() => Cancelled = false;
+    }
+
+    [ByRefEvent]
+    public sealed class GotEntityChatTransformEvent
+    {
+        public readonly EntityUid Recipient;
+        public EntityChat Chat;
+        public EntityChatData RecipientData;
+
+        public GotEntityChatTransformEvent(EntityUid recipient, EntityChat chat, EntityChatData recipientData)
+        {
+            Recipient = recipient;
+            Chat = chat;
+            RecipientData = recipientData;
+        }
+
+        public bool Handled;
+    }
+
+    [ByRefEvent]
+    public sealed class GotEntityChatEvent
```

Comment:
Using a class byref does nothing.
```suggestion
    [ByRefEvent]
    public record struct GotEntityChatEvent
```